README.md
# Buffet

*All-inclusive Buffer for C*  

![schema](assets/buffet.png)  

Experimental string buffer featuring

- **SSO** (small string optimization) inline short data into the type.
- **views** (or 'slices') no-copy references to sub-strings.
- **refcount** (reference counting) account for views before mutation or release.
- automated allocations

Aims at [**security**](#Security) with decent [**speed**](#Bench).  
Todo: thread safety


[**API**](#API)  

![CI](https://github.com/alcover/buffet/actions/workflows/ci.yml/badge.svg)

## Layout

Buffet is a tagged union with 4 modes.  

```C
// Hard values show 64-bit

union Buffet {
    struct ptr {
        char*   data
        size_t  len
        size_t  off:62, tag:2 // tag = OWN|SSV|VUE
    }
    struct sso {
        char    data[22]
        uint8_t refcnt
        uint8_t len:6, tag:2  // tag = SSO
    }
}

sizeof(Buffet) == 24
```
The *tag* sets a Buffet's mode :  

- `OWN`  co-owning slice of a store
- `SSO`  embedded char array
- `SSV`  (small string view) view on an SSO
- `VUE`  non-owning view on any data

If OWN, *Buffet.data* points into an allocated heap store :

```C
struct Store {
    size_t   cap    // store capacity
    size_t   len    // store length
    uint32_t refcnt // number of views on store
    uint32_t canary // invalidates store if modified
    char     data[] // buffer data, shared by owning views
}
```


#### Schema

![schema](assets/schema.png)

```C
#include "../buffet.h"

int main() {
    
// SHARED OWN =================

    char large[] = "DATA STORE IS HEAP ALLOCATION.";

    Buffet own1 = bft_memcopy(large, sizeof(large)-1);
    // Now own1 owns a store housing a copy of `large`
    bft_dbg(&own1); 
    //-> OWN 30 "DATA STORE ..."
    // View "STORE" in own1 :
    Buffet own2 = bft_view(&own1, 5, 5);
    // Now own1 and own2 share the store, whose refcount is 2
    bft_dbg(&own2); 
    //-> OWN 5 "STORE"

// SSO & SSV =================

    char small[] = "SMALL STRING";

    Buffet sso1 = bft_memcopy(small, sizeof(small)-1);
    bft_dbg(&sso1); 
    //-> SSO 12 "SMALL STRING"
    // View "STRING" in sso1 :
    Buffet ssv1 = bft_view(&sso1, 6, 6);
    bft_dbg(&ssv1); 
    //-> SSV 6 "STRING"

// VUE =======================

    char any[] = "SOME BYTES";

    // View "BYTES" in `any` :
    Buffet vue1 = bft_memview(any+5, 5);
    bft_dbg(&vue1); 
    //-> VUE 5 "BYTES"

    return 0;
}
```


### Build & check

    make && make check

While extensive, unit tests may not yet cover all cases.


### Security

Buffet aims at preventing memory faults, including from user.  
(Except of course losing scope and such.)  

```C
// (pseudo code)

// overflow
Buffet buf = bft_new(8)
bft_append(buf, large_str) // Done

// invalid ref
Buffet buf = bft_memcopy(short_str)
Buffet view = bft_view(buf)
bft_append(buf, large_str) // would mutate SSO buf into OWN
// Abort + warning "Append would invalidate views on SSO"

// double-free
bft_free(buf)
bft_free(buf) // OK

// use-after-free
bft_free(buf)
bft_append(buf, "foo") // Done. Now buf is "foo".

// aliasing
Buffet alias = buf // should be `alias = bft_dup(buf)`
bft_free(buf)
bft_free(alias) // OK. Possible warning "Bad canary. Double free ?"

// Etc...
```

To this end, operations like *view* or *free* may check the store's canary and refcount.  
If wrong, the operation aborts and returns an empty buffet.  

See *src/check.c* unit-tests and warnings output.


### Bench

`make && make bench` (requires *libbenchmark-dev*)  

NB: The lib is not much optimized and the bench maybe amateurish.  

On a weak Core i3 :  
<pre>
MEMVIEW_cpp/8                1 ns          1 ns 1000000000
MEMVIEW_buffet/8             7 ns          7 ns  103101477
MEMCOPY_c/8                 17 ns         17 ns   41771438
MEMCOPY_buffet/8            12 ns         12 ns   59136008
MEMCOPY_c/32                16 ns         16 ns   44280093
MEMCOPY_buffet/32           27 ns         27 ns   25569904
MEMCOPY_c/128               17 ns         17 ns   41047431
MEMCOPY_buffet/128          31 ns         31 ns   22837310
MEMCOPY_c/512               28 ns         28 ns   25499520
MEMCOPY_buffet/512          45 ns         45 ns   15477668
MEMCOPY_c/2048              94 ns         94 ns    7401330
MEMCOPY_buffet/2048        110 ns        110 ns    6405395
MEMCOPY_c/8192             199 ns        199 ns    3509839
MEMCOPY_buffet/8192        284 ns        284 ns    2461398
APPEND_cpp/8/4              12 ns         12 ns   57633324
APPEND_buffet/8/4           22 ns         22 ns   32506683
APPEND_cpp/8/16             32 ns         32 ns   21765849
APPEND_buffet/8/16          30 ns         30 ns   23278437
APPEND_cpp/24/4             48 ns         48 ns   14586050
APPEND_buffet/24/4          31 ns         31 ns   22807349
APPEND_cpp/24/32            47 ns         47 ns   14982967
APPEND_buffet/24/32         29 ns         29 ns   24023279
SPLITJOIN_c               2913 ns       2913 ns     240557
SPLITJOIN_cpp             3128 ns       3128 ns     223070
SPLITJOIN_buffet          1431 ns       1431 ns     487453
</pre>


# API

[bft_new](#bft_new)  
[bft_memcopy](#bft_memcopy)  
[bft_memview](#bft_memview)  
[bft_copy](#bft_copy)  
[bft_copyall](#bft_copyall)  
[bft_view](#bft_view)  
[bft_dup](#bft_dup)  (don't alias buffets, use this)  
[bft_append](#bft_append)  
[bft_split](#bft_split)  
[bft_splitstr](#bft_splitstr)  
[bft_join](#bft_join)  
[bft_free](#bft_free)  

[bft_cmp](#bft_cmp)  
[bft_cap](#bft_cap)  
[bft_len](#bft_len)  
[bft_data](#bft_data)  
[bft_cstr](#bft_cstr)  
[bft_export](#bft_export)  

[bft_print](#bft_print)  
[bft_dbg](#bft_dbg)  


### bft_new

    Buffet bft_new (size_t cap)

Create a new empty Buffet of minimum capacity *cap*.  

```C
Buffet buf = bft_new(40);
bft_dbg(&buf); 
// OWN 0 ""
```

### bft_memcopy

    Buffet bft_memcopy (const char *src, size_t len)

Create a new Buffet by copying *len* bytes from *src*.  

```C
Buffet copy = bft_memcopy("Bonjour", 3);
// SSO 3 "Bon"
```

### bft_memview

    Buffet bft_memview (const char *src, size_t len)

Create a new Buffet viewing *len* bytes from *src*.  
You get a window into *src* without copy or allocation.  
NB: You shouldn't directly *memview* a Buffet's data. Use *view()*

```C
char src[] = "Eat Buffet!";
Buffet view = bft_memview(src+4, 6);
// VUE 6 "Buffet"
```

### bft_copy

    Buffet bft_copy (const Buffet *src, ptrdiff_t off, size_t len)

Copy *len* bytes at offset *off* from Buffet *src* into a new Buffet.  

```C
Buffet src = bft_memcopy("Bonjour", 7);
Buffet cpy = bft_copy(&src, 3, 4);
// SSO 4 "jour"
```


### bft_copyall

    Buffet bft_copyall (const Buffet *src)

Copy all bytes from Buffet *src* into a new Buffet.


### bft_view

    Buffet bft_view (Buffet *src, ptrdiff_t off, size_t len)

View *len* bytes of Buffet *src*, starting at *off*.  
You get a window into *src* without copy or allocation.  

The return internal type depends on *src* type : 

- `view(SSO) -> SSV` (refcounted)
- `view(SSV) -> SSV` on *src*'s target
- `view(OWN) -> OWN` (as refcounted store co-owner)
- `view(VUE) -> VUE` on *src*'s target

If the return is OWN, the target store won't be released before either  
- the return is discarded by *bft_free*
- the return is detached by e.g. appending to it.

```C
#include "../buffet.h"

int main() {

    char text[] = "Bonjour monsieur Buddy. Already speaks french!";

    // view sso
    Buffet sso = bft_memcopy(text, 16); // "Bonjour monsieur"
    Buffet ssv = bft_view(&sso, 0, 7);
    bft_dbg(&ssv);

    // view ssv
    Buffet Bon = bft_view(&ssv, 0, 3);
    bft_dbg(&Bon);

    // view own
    Buffet own = bft_memcopy(text, sizeof(text));
    Buffet ownview = bft_view(&own, 0, 7);
    bft_dbg(&ownview);

    // detach view
    bft_append (&ownview, "!", 1);
    // bft_free(&ownview); 
    bft_free(&own); // Done

    // view vue
    Buffet vue = bft_memview(text+8, 8); // "Good"
    Buffet mon = bft_view(&vue, 0, 3);
    bft_dbg(&mon);

    return 0;
}
```

```
$ cc view.c libbuffet.a -o view && ./view
SSV 7 data:"Bonjour"
SSV 3 data:"Bon"
OWN 7 data:"Bonjour"
VUE 3 data:"mon"
```

### bft_dup

    Buffet bft_dup (const Buffet *src)

Create a shallow copy of *src*.  
**Use this intead of aliasing a Buffet.**  

```C
Buffet src = bft_memcopy("Hello", 5);
Buffet cpy = src; // BAD
Buffet cpy = bft_dup(&src); // GOOD
bft_dbg(&cpy);
// SSO 5 "Hello"
```

Rem: aliasing would mostly work but mess up refcounting (without crash if store protections are enabled) :  

```C
Buffet alias = sso; //ok if sso was not viewed
Buffet alias = own; //not refcounted
Buffet alias = vue; //ok
```

### bft_free

    void bft_free (Buffet *buf)

Discards *buf*.  

- aborts if buf is an SSO with views.
- otherwise, buf is zeroed into an empty SSO.
- if buf was a view, its target refcount is decremented.
- if buf was the last view on a store, the store is released.  

Security:
- the zeroing makes double-free harmless.
- the only problematic use-after-free would be of a OWN alias (not recommended), but the store management prevents stale memory access.

```C
#include "../buffet.h"

int main() {

    char text[] = "Le grand orchestre de Patato Valdez";

    Buffet own = bft_memcopy(text, sizeof(text));
    Buffet ref = bft_view(&own, 9, 9); // "orchestre"
    bft_free(&own); // A bit soon but ok, --refcnt
    bft_dbg(&own);  // SSO 0 ""
    bft_free(&ref); // Was last co-owner, store is released

    Buffet sso = bft_memcopy(text, 8); // "Le grand"
    Buffet ref2 = bft_view(&sso, 3, 5); // "grand"
    bft_free(&sso); // WARN line:328 bft_free: SSO has views on it
    bft_free(&ref2);
    bft_free(&sso); // OK now
    bft_dbg(&sso);  // SSO 0 ""

    return 0;
}
 ```

```
$ valgrind  --leak-check=full ./bin/ex/free
All heap blocks were freed -- no leaks are possible
```

### bft_cat

    size_t bft_cat (Buffet *dst, const Buffet *buf, const char *src, size_t len)

Concatenates *buf* and *len* bytes of *src* into resulting *dst*.  
Returns total length or 0 on error.

```C
Buffet buf = bft_memcopy("abc", 3);
Buffet dst;
size_t totlen = bft_cat(&dst, &buf, "def", 3);
bft_dbg(&dst);
// SSO 6 "abcdef"
```

### bft_append

    size_t bft_append (Buffet *dst, const char *src, size_t len)

Appends *len* bytes from *src* to *dst*.  
Returns new length or 0 on error.

```C
Buffet buf = bft_memcopy("abc", 3); 
size_t newlen = bft_append(&buf, "def", 3);
bft_dbg(&buf);
// SSO 6 "abcdef"
```

NB: returns failure if *buf* has views and would mutate from SSO to OWN to increase capacity, invalidating the views :  

```C
Buffet foo = bft_memcopy("short foo ", 10);
Buffet view = bft_view(&foo, 0, 5);
// would mutate to OWN :
size_t rc = bft_append(&foo, "now too long for SSO");
assert(rc==0); // meaning aborted
```

To prevent this, release views before appending to a small buffet.  

### bft_split

    Buffet* bft_split (const char* src, size_t srclen, const char* sep, size_t seplen, 
    int *outcnt)

Splits *src* along separator *sep* into a Buffet Vue list of length `*outcnt`.  

Being made of views, you can `free(list)` without leak provided no element was made an owner by e.g appending to it.

### bft_splitstr


ChangeLog
2022-07-06  alco

	v0.7.3

	* void free()
	* copy_all()
	* Store.len vs Store.end
	* new append() & cat() implementations
	* doxygen headers


2022-06-20  alco

	v0.7.2 - SSV tag: view on SSO

	Simplify viewing an SSO thanks to a dedicated tag

2022-06-15  alco

	v0.7.1 - SSO refcnt

	Before, taking a view of an SSO would auto-mutate it into OWN with alloc.
	Beyond the cost of this convertion, it created a fault if a VUE was active 
	on the SSO..
	Now taking a view of an SSO uses new_ssovue() which increases the new
	SSO.refcnt field and uses a non-zero VUE.off as flag and offset

	* buffet_cat
	* buffet_copyall

2022-06-09  alco

	v0.7.0 - Shared ownership

	* remove 'REF' tag
	* BuffetPtr.aux becomes BuffetPtr.off, used as OWN's offset in Store
	* new Store layout : add .cap and .end fields
	* getstore() + new_store()
	* remove grow()

2022-05-19  alco
	
	v0.6.1
	
	* reset Buffet by memset

	* buffet_view: canary check for onwner only

2022-05-04  alco
	
	v0.6.0
	
	* NEW LAYOUT: use size_t capacity instead of hard-coded bit-length.
	This makes Buffet a 24 bytes type on 64 bit.

	* introduce micro-benchmarking for future fine-tuning

2022-04-20  alco
	
	v0.5.2
	
	* buffet_free: only return false on a valid owner with live views

	* validate Store on every access

2022-04-20  alco
	
	v0.5.1
	
	* buffet_clone

	* buffet_append: solving self-append 
	(as pointed by Justas Masiulis, if source is Buffet itself and relocates, 
	we copy from invalid memory)

	* bench: fast append_cpp
	
	* OVERALLOC int

src/check.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include "buffet.h"
#include "log.h"
#include "util.h"

#ifdef NDEBUG
#undef NDEBUG
#endif

#define alphalen (128)
char alpha[alphalen+1];
char tmp[alphalen+1];

// put a slice of alpha into tmp
static char* take (size_t off, size_t len) {
    assert(off+len <= alphalen);
    memcpy(tmp, alpha+off, len);
    tmp[len] = 0;
    return tmp;
}

#define assert_int(val, exp) { if ((int)(val) != (int)(exp)) { \
fprintf(stderr, "%d: %s:%d != %d\n", __LINE__, #val, (int)(val), (int)(exp)); \
exit(EXIT_FAILURE);}}

#define assert_str(val, exp) { if (strcmp((val), (exp))) {\
fprintf(stderr, "%d: %s:'%s' != '%s'\n", __LINE__, #val, (val), (exp)); \
exit(EXIT_FAILURE);}}

#define assert_stn(val, exp, n) { if (strncmp((val), (exp), (n))) {\
fprintf(stderr, \
    "%d: %d bytes %s:'%.*s' != '%s'\n", __LINE__, (int)(n), #val, (int)(n), (val), (exp)); \
exit(EXIT_FAILURE);}}


#define check_props(buf, off, len) do{ \
    /* len */ \
    assert_int (bft_len(buf), len); \
    /* data */ \
    const char *exp = take(off,len); \
    assert_stn (bft_data(buf), exp, len); \
    /* cstr */ \
    bool mustfree; \
    const char *cstr = bft_cstr(buf, &mustfree); \
    assert_str (cstr, exp); \
    if (mustfree) free((char*)cstr); \
    /* export */ \
    char *export = bft_export(buf); \
    assert_str (export, exp); \
    free(export); \
}while(0)

#define check_zero(buf) { \
    assert_int (bft_len(buf), 0); \
    assert_str (bft_data(buf), ""); \
    bool mustfree; \
    const char *cstr = bft_cstr(buf, &mustfree); \
    assert_str (cstr, ""); \
    assert(!mustfree); \
}

#define around(fun, n) fun((n)-1); fun(n); fun((n)+1);

#define serie(fun, off) \
    fun (off, 0); \
    fun (off, 1); \
    fun (off, 8); \
    fun (off, BUFFET_SSOMAX-1); \
    fun (off, BUFFET_SSOMAX); \
    fun (off, BUFFET_SSOMAX+1); \
    fun (off, 24); \
    fun (off, 32); \
    fun (off, 48); \
    fun (off, 64); 


//=============================================================================
void unew (size_t cap) {
    Buffet buf = bft_new(cap); 
    check_props(&buf, 0, 0);
    bft_free(&buf); 
}

void new() 
{ 
    unew (0);
    unew (1);
    unew (8);
    around (unew, BUFFET_SSOMAX);
    around (unew, sizeof(Buffet));
    around (unew, 32);
    around (unew, 64);
    around (unew, 1024);
    around (unew, 4096);
}

//=============================================================================
void umemcopy(size_t off, size_t len) {
    Buffet buf = bft_memcopy(alpha+off, len);
    check_props(&buf, off, len);    
    bft_free(&buf);
}

void memcopy() 
{
    serie(umemcopy, 0);
    serie(umemcopy, 8);
}

//=============================================================================
void umemview (size_t off, size_t len) {
    Buffet buf = bft_memview (alpha+off, len);
    check_props(&buf, off, len);
    bft_free(&buf);
}

void memview() 
{
    serie(umemview, 0);
    serie(umemview, 8);
}

//=============================================================================
#define dup_new(len) { \
    Buffet src = bft_new(len); \
    Buffet buf = bft_dup(&src); \
    check_props(&buf, 0, 0); \
    bft_free(&buf); \
    bft_free(&src); \
}

#define dup_memcopy(len) { \
    Buffet src = bft_memcopy(alpha, len); \
    Buffet buf = bft_dup(&src); \
    check_props(&buf, 0, len); \
    bft_free(&buf); \
    bft_free(&src); \
}

#define dup_memview(len) { \
    Buffet src = bft_memview(alpha, len); \
    Buffet buf = bft_dup(&src); \
    check_props(&buf, 0, len); \
    bft_free(&buf); \
    bft_free(&src); \
}

#define dup_view(srclen, len) { \
    Buffet src = bft_memcopy(alpha, srclen); \
    Buffet ref = bft_view(&src, 0, len); \
    Buffet buf = bft_dup(&ref); \
    check_props(&buf, 0, len); \
    bft_free(&buf); \
    bft_free(&ref); \
    bft_free(&src); \
}

void dup()
{
    dup_new(0);
    dup_new(8);
    dup_new(32);
    dup_memcopy(0);
    dup_memcopy(8);
    dup_memcopy(32);
    dup_memview(0);
    dup_memview(8);
    dup_memview(32);
    dup_view(8, 0);
    dup_view(8, 4);
    dup_view(32, 0);
    dup_view(32, 16);
}


//=============================================================================
void ucopy (size_t off, size_t len) {
    Buffet src = bft_memcopy(alpha, alphalen);
    Buffet buf = bft_copy(&src, off, len);
    check_props(&buf, off, len);
    bft_free(&buf);
    bft_free(&src);
}

void copy() 
{
    serie(ucopy, 0);
    serie(ucopy, 8);
    ucopy (0, alphalen);
}


//===== VIEW ==================================================================

#define viewcheck(src, srclen, off, len) \
    Buffet view = bft_view(&src, off, len); \
    size_t explen = len; \
    if (off>=srclen) explen = 0; \
    else if (off+len>=srclen) explen = srclen-off; \
    assert_int (bft_len(&view), explen); \
    assert_stn (bft_data(&view), alpha+off, explen); \
    bft_free(&view); \
    bft_free(&src);


#define view_new(cap, off, len) { \
    Buffet src = bft_new(cap); \
    Buffet view = bft_view(&src, off, len); \
    assert_int (bft_len(&view), 0); \
    assert_str (bft_data(&view), ""); \
}

#define view_memcopy(srclen, off, len) { \
    Buffet src = bft_memcopy(alpha, srclen); \
    viewcheck(src, srclen, off, len) \
}

#define view_memview(srclen, off, len) { \
    Buffet src = bft_memview (alpha, srclen); \
    viewcheck(src, srclen, off, len) \
}

#define view_view(srclen, off, len) { \
    Buffet buf = bft_memcopy(alpha, alphalen); \
    Buffet src = bft_view (&buf, 0, srclen); \
    viewcheck(src, srclen, off, len) \
    bft_free(&buf); \
}

#define VIEWCASES(fun, srclen)\
fun (srclen, 0, 0); \
fun (srclen, 0, srclen/2);  /* part */ \
fun (srclen, 0, srclen);    /* full */ \
fun (srclen, 0, srclen+1);  /* beyond */ \
fun (srclen, 2, srclen/2);  /* part */ \
fun (srclen, 2, srclen-2);  /* till end */ \
fun (srclen, 2, srclen+1);  /* beyond */ \
fun (srclen, srclen, 0);    /* bad off */ \
fun (srclen, srclen, 1);    /* bad off */ \
fun (srclen, srclen+1, 0);  /* bad off */ \
fun (srclen, srclen+1, 1);  /* bad off */ \

#define view_alias_after_free(initlen) { \
    Buffet src = bft_memcopy(alpha, initlen); \
    Buffet alias = src; \
    bft_free(&src);     \
    Buffet ref = bft_view (&alias, 0, initlen); \
    check_props(&ref, 0, 0); \
    bft_free(&ref);  \
}

void view() 
{
    view_new (0, 0, 0);
    view_new (0, 8, 0);
    view_new (0, 8, 8);
    view_new (8, 0, 0);
    view_new (8, 0, 4);
    view_new (8, 0, 8);
    view_new (32, 0, 0);
    view_new (32, 0, 16);
    view_new (32, 0, 32);

    VIEWCASES (view_memcopy, 8)
    VIEWCASES (view_memcopy, 32)
    VIEWCASES (view_memview, 8)
    VIEWCASES (view_memview, 32)
    VIEWCASES (view_view, 8)
    VIEWCASES (view_view, 32)
    
    // view_alias_after_free(8); // useless. SSO alias is copy.
    view_alias_after_free(32);
}

//==============================================================================

#define ucat(dst, buf, buflen, len) {\
    size_t rc = bft_cat(dst, buf, alpha+buflen, len); \
    size_t explen = buflen+len; \
    assert_int(rc, explen); \
    check_props(dst, 0, explen);\
    bft_free(dst); \
    bft_free(buf); \
}

#define cat_new(cap, len) {\
    Buffet dst; \
    Buffet buf = bft_new(cap);\
    ucat (&dst, &buf, 0, len) \
}

#define cat_init(op, buflen, len) {\
    Buffet dst; \
    Buffet buf = bft_##op(alpha, buflen); \
    ucat (&dst, &buf, buflen, len); \
}

#define cat_view(buflen, len) { \
    Buffet src = bft_memcopy(alpha, buflen); \
    Buffet ref = bft_view (&src, 0, buflen); \
    Buffet dst; \
    ucat (&dst, &ref, buflen, len) \
    bft_free(&src); \
}

void cat()
{
    cat_new (0, 0);
    cat_new (0, 4);
    cat_new (0, 32);
    cat_new (4, 0);
    cat_new (4, 4);
    cat_new (4, 32);
    cat_new (32, 0);    
    cat_new (32, 4);
    cat_new (32, 32);

    cat_init (memcopy, 0, 0);
    cat_init (memcopy, 0, 4);
    cat_init (memcopy, 0, 32);
    cat_init (memcopy, 4, 0);
    cat_init (memcopy, 4, 4);
    cat_init (memcopy, 4, BUFFET_SSOMAX-4);
    cat_init (memcopy, 4, BUFFET_SSOMAX-3);
    cat_init (memcopy, 4, 32);
    cat_init (memcopy, 32, 0);
    cat_init (memcopy, 32, 32);

    cat_init (memview, 0, 0);
    cat_init (memview, 0, 4);
    cat_init (memview, 0, 32);
    cat_init (memview, 4, 0);
    cat_init (memview, 4, 4);
    cat_init (memview, 4, BUFFET_SSOMAX-4);
    cat_init (memview, 4, BUFFET_SSOMAX-3);
    cat_init (memview, 4, 32);
    cat_init (memview, 32, 0);
    cat_init (memview, 32, 32);
    
    cat_view (0, 0);
    cat_view (0, 4);
    cat_view (0, 32);
    cat_view (4, 0);
    cat_view (4, 4);
    cat_view (4, 4);
    cat_view (4, BUFFET_SSOMAX-4);
    cat_view (4, BUFFET_SSOMAX-3);
    cat_view (4, 32);
    cat_view (32, 0);
    cat_view (32, 32);
}

//==============================================================================

void uapn(Buffet *buf, size_t buflen, size_t len) {
    size_t rc = bft_append (buf, alpha+buflen, len); 
    size_t explen = buflen+len; 
    assert_int(rc, explen); 
    check_props(buf, 0, explen); 
    bft_free(buf); 
}

#define apn_new(cap, len) {\
    Buffet buf = bft_new(cap);\
    uapn (&buf, 0, len); \
}

#define apn_init(op, buflen, len) {\
    Buffet buf = bft_##op (alpha, buflen); \
    uapn (&buf, buflen, len); \
}

#define apn_view(buflen, len) { \
    Buffet src = bft_memcopy(alpha, buflen); \
    Buffet ref = bft_view (&src, 0, buflen); \
    uapn (&ref, buflen, len); \
    bft_free(&src); \
}

#define apn_viewed(buflen, len, exprc) { \
    Buffet buf = bft_memcopy(alpha, buflen); \
    Buffet ref = bft_view (&buf, 0, buflen); \
    /*bft_dbg(&buf);*/ \
    size_t rc = bft_append (&buf, alpha+buflen, len); \
    assert_int(rc, exprc); \
    if (rc) check_props(&buf, 0, (buflen+len)); \
    bft_free(&ref); \
    bft_free(&buf); \
}

void append()
{
    apn_new (0, 0);
    apn_new (0, 4);
    apn_new (0, 32);
    apn_new (4, 0);
    apn_new (4, 4);
    apn_new (4, 32);
    apn_new (32, 0);    
    apn_new (32, 4);
    apn_new (32, 32);

    apn_init (memcopy, 0, 0);
    apn_init (memcopy, 0, 4);
    apn_init (memcopy, 0, 32);
    apn_init (memcopy, 4, 0);
    apn_init (memcopy, 4, 4);
    apn_init (memcopy, 4, BUFFET_SSOMAX-4);
    apn_init (memcopy, 4, BUFFET_SSOMAX-3);
    apn_init (memcopy, 4, 32);
    apn_init (memcopy, 32, 0);
    apn_init (memcopy, 32, 32);
    
    apn_init (memview, 0, 0);
    apn_init (memview, 0, 4);
    apn_init (memview, 0, 32);
    apn_init (memview, 4, 0);
    apn_init (memview, 4, 4);
    apn_init (memview, 4, BUFFET_SSOMAX-4);
    apn_init (memview, 4, BUFFET_SSOMAX-3);
    apn_init (memview, 4, 32);
    apn_init (memview, 32, 0);
    apn_init (memview, 32, 32);
    
    apn_view (0, 0);
    apn_view (0, 4);
    apn_view (0, 32);
    apn_view (4, 0);
    apn_view (4, 4);
    apn_view (4, 4);
    apn_view (4, BUFFET_SSOMAX-4);
    apn_view (4, BUFFET_SSOMAX-3);
    apn_view (4, 32);
    apn_view (32, 0);
    apn_view (32, 32);

    apn_viewed (8, 4, 12);
    apn_viewed (8, 32, 0); // would mutate
    apn_viewed (BUFFET_SSOMAX+1, 32, (BUFFET_SSOMAX+1+32));
}


//==============================================================================

#define usploin(src, sep) { \
    size_t srclen = strlen(src);\
    size_t seplen = strlen(sep);\
    int cnt; \
    Buffet* parts = bft_split (src, srclen, sep, seplen, &cnt); \
    Buffet joined = bft_join (parts, cnt, sep, seplen); \
    assert_str (bft_data(&joined), src); \
    assert_int (bft_len(&joined), srclen); \
    free(parts);\
    bft_free(&joined);\
}

#define sploin(a,b,sep) \
    usploin ("", #sep); \
    usploin (#sep, #sep); \
    usploin (#a, #sep); \
    usploin (#a #sep,  #sep);  \
    usploin (#a #sep #b,  #sep);  \
    usploin (#a #sep #b #sep,  #sep);  \
    usploin (#sep #a,  #sep); \
    usploin (#sep #a #sep,  #sep);  \
    usploin (#sep #a #sep #b,  #sep);  \
    usploin (#sep #a #sep #b #sep,  #sep);  \
    usploin (#a #sep #sep,  #sep);  \
    usploin (#a #sep #sep #b,  #sep);  \
    usploin (#a #sep #sep #b #sep #sep,  #sep);  \
    usploin (#sep #sep #a,  #sep); \
    usploin (#sep #sep #a #sep #sep,  #sep);  \
    usploin (#sep #sep #a #sep #sep #b,  #sep);  \
    usploin (#sep #sep #a #sep #sep #b #sep #sep, #sep); 

void splitjoin() 
{ 
    sploin (a, b, |)
    sploin (a, b, ||)
    sploin (foo, bar, |)
    sploin (foo, bar, ||)
}

//=============================================================================

#define check_free(buf) {\
    bft_free(buf); \
    check_zero(buf); \
}

#define free_new(len) { \
    Buffet buf = bft_new (len); \
    check_free(&buf); \
}
#define free_memcopy(len) { \
    Buffet buf = bft_memcopy(alpha, len); \
    check_free(&buf); \
}
#define free_memview(len) { \
    Buffet buf = bft_memview (alpha, len); \
    check_free(&buf); \
}
#define free_view(len) { \
    Buffet own = bft_memcopy(alpha, 32); \
    Buffet ref = bft_view (&own, 0, len); \
    check_free(&ref); \
    check_free(&own); \
}
#define free_copy(len) { \
    Buffet own = bft_memcopy(alpha, 32); \
    Buffet cpy = bft_copy (&own, 0, len); \
    check_free(&cpy); \
    check_free(&own); \
}

#define double_free(len) { \
    Buffet buf = bft_memcopy(alpha, len); \
    check_free(&buf); \
    check_free(&buf); \
}
#define double_free_ref(srclen, len) { \
    Buffet src = bft_memcopy(alpha, srclen); \
    Buffet ref = bft_view (&src, 0, len); \
    check_free(&ref); \
    check_free(&ref); \
    check_free(&src); \
}
#define free_alias(len) { \
    Buffet src = bft_memcopy(alpha, len); \
    Buffet alias = src; \
    check_free(&src); \
    check_free(&alias); \
}
#define free_ref_alias(len) { \
    Buffet own = bft_memcopy(alpha, 32); \
    Buffet ref = bft_view (&own, 0, len); \
    Buffet alias = ref; \
    check_free(&ref); \
    check_free(&alias); \
    check_free(&own); \
}
#define free_viewed(len, intact) { \
    Buffet src = bft_memcopy(alpha, len); \
    Buffet ref = bft_view (&src, 0, len); \
    bft_free(&src); \
    if(intact) check_props(&src, 0, len); \
    else check_zero(&src); \
    check_free(&ref); \
}


void free_()
{
    free_new(0)
    free_new(8)
    free_new(32)
    free_memcopy(0)
    free_memcopy(8)
    free_memcopy(32)
    free_memview(0)
    free_memview(8)
    free_memview(32)
    free_copy(0)
    free_copy(8)
    free_copy(32)
    free_view(0)
    free_view(8)
    free_view(32)

    /**** danger ****/

    double_free(0);
    double_free(8);
    double_free(32);

    double_free_ref(8, 4);
    double_free_ref(32, 8);
    double_free_ref(32, 20);

    free_alias(8);
    free_alias(32);
    
    free_viewed (0, true)
    free_viewed (8, true)
    free_viewed (32, false)

    free_ref_alias (0)
    free_ref_alias (8)
    free_ref_alias (32)
}

//=============================================================================

#define assert_cmp(a, b, exp) \
int cmp = bft_cmp(a, b); \
if (cmp) cmp = cmp/abs(cmp); /*exp = abs(exp);*/ \
assert_int(cmp, exp);


#define ucmp_memop(op, lena, lenb, exp) { \
    Buffet a = bft_##op(alpha, lena); \
    Buffet b = bft_##op(alpha, lenb); \
    assert_cmp(&a, &b, exp); \
    bft_free(&a); \
    bft_free(&b); \
}

#define ucmp_view(len, exp) { \
    Buffet buf = bft_memcopy(alpha, len); \
    Buffet view = bft_view(&buf, 0, len); \
    assert_int (!!bft_cmp(&view, &buf), !!exp); \
    bft_free(&view); \
    bft_free(&buf); \
}

void cmp()
{
    ucmp_memop(memcopy, 0, 0, 0);
    ucmp_memop(memcopy, 1, 1, 0);
    ucmp_memop(memcopy, 8, 8, 0);
    ucmp_memop(memcopy, 32, 32, 0);
    ucmp_memop(memcopy, 1, 2, -1);
    ucmp_memop(memcopy, 8, 9, -1);
    ucmp_memop(memcopy, 32, 33, -1); 

    ucmp_memop(memview, 0, 0, 0);
    ucmp_memop(memview, 1, 1, 0);
    ucmp_memop(memview, 8, 8, 0);
    ucmp_memop(memview, 32, 32, 0);
    ucmp_memop(memview, 1, 2, -1);
    ucmp_memop(memview, 8, 9, -1);
    ucmp_memop(memview, 32, 33, -1);

    ucmp_view(0, 0);
    ucmp_view(1, 0);
    ucmp_view(8, 0);
    ucmp_view(32, 0);

    // todo other combins
}

//=============================================================================
void zero()
{
    Buffet buf = BUFFET_ZERO;

    assert (!buf.ptr.data);
    assert (!buf.ptr.len);
    assert (!buf.ptr.off);
    assert (!buf.ptr.tag);

    assert (!strcmp(buf.sso.data, ""));
    assert (!buf.sso.len);
    assert (!buf.sso.refcnt);
    assert (!buf.sso.tag); // better cmp to SSO tag.. make it public ?
}

//=============================================================================

#define run(name) \
LOG("> " #name); \
name(); \
LOG("%s OK", #name); 

int main()
{
    repeatat(alpha, alphalen, ALPHA64);
    
    LOG("unit tests... ");
    run(zero);
    run(new);
    run(memcopy);
    run(memview);
    run(dup);
    run(copy);
    run(view);
    run(cat);
    run(append);
    run(splitjoin);
    run(free_);
    run(cmp);
    LOG("unit tests OK");

    return 0;
}